File: 0001~0500/0031. Next Permutation/main.py
"""
Source: https://rb.gy/93w50
Date: 2023/8/1
Skill:
Ref:
Runtime: 41 ms, faster than 98.97%
Memory Usage: 16.15 MB, less than 95.86%
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        idx, idx2 = len(nums) - 1, len(nums) - 1
        while idx - 1 >= 0 and nums[idx] <= nums[idx - 1]:
            idx -= 1
        if idx == 0:
            nums.sort()
            return
        idx -= 1
        while nums[idx2] <= nums[idx]:
            idx2 -= 1
        nums[idx], nums[idx2] = nums[idx2], nums[idx]
        nums[idx + 1:] = sorted(nums[idx + 1:])


if __name__ == "__main__":
    s = Solution()

File: 0001~0500/0031. Next Permutation/main.cpp
/**
 * Source: ibit.ly/11lhZ
 * Date: 2023/11/17
 * Skill:
 * Runtime: 4 ms, faster than 49.55% of C++ online submissions
 * Memory Usage: 12.37 MB, less than 91.28% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */


#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>


using namespace std;


class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n = nums.size(), pos = nums.size() - 1;
        while (pos > 0 && nums[pos] <= nums[pos - 1]) pos--;
        if (pos == 0) {
            int end = n - 1;
            while (pos < end) swap(nums[pos++], nums[end--]);
        } else {
            int a = n - 1;
            while (nums[a] <= nums[pos - 1]) a--;
            swap(nums[a], nums[pos - 1]);
            a = n - 1;
            while (pos < a) swap(nums[pos++], nums[a--]);
        }
    }
};


int main() {
    Solution s;
}